dpcpp/matrix/dense_kernels.dp.cpp

```diff
@@ -65,11 +68,926 @@ namespace dpcpp {
+                                           tmp_work_array, reduce_op);
+
+    if (local_id == 0) {
+        work[thread::get_block_id(item_ct1)] = tmp_work_array[0];
+    }
+}
+
+
+template <std::uint32_t cfg = KCFG_1D::encode(256, 32), typename ValueType,
+          typename CallableReduce, typename CallableFinalize>
+void finalize_reduce_computation(
+    size_type size, const ValueType *work, ValueType *result,
+    CallableReduce reduce_op, CallableFinalize finalize_op,
+    sycl::nd_item<3> item_ct1,
+    UninitializedArray<ValueType, KCFG_1D::decode<0>(cfg)> *tmp_work)
+{
+    constexpr auto wg_size = KCFG_1D::decode<0>(cfg);
+    constexpr auto sg_size = KCFG_1D::decode<1>(cfg);
+
+    const auto local_id = thread::get_local_thread_id<sg_size>(item_ct1);
+
+    ValueType tmp = zero<ValueType>();
+    for (auto i = local_id; i < size; i += wg_size) {
+        tmp = reduce_op(tmp, work[i]);
+    }
+    ValueType *tmp_work_array = *tmp_work;
+    tmp_work_array[local_id] = tmp;
+
+    ::gko::kernels::dpcpp::reduce<sg_size>(group::this_thread_block(item_ct1),
+                                           tmp_work_array, reduce_op);
+
+    if (local_id == 0) {
+        *result = finalize_op(tmp_work_array[0]);
+    }
+}
+
+
+template <std::uint32_t cfg = KCFG_1D::encode(256, 32), typename ValueType>
+void compute_partial_dot(
+    size_type num_rows, const ValueType *__restrict__ x, size_type stride_x,
+    const ValueType *__restrict__ y, size_type stride_y,
+    ValueType *__restrict__ work, sycl::nd_item<3> item_ct1,
+    UninitializedArray<ValueType, KCFG_1D::decode<0>(cfg)> *tmp_work)
+{
+    compute_partial_reduce<cfg>(
+        num_rows, work,
+        [x, stride_x, y, stride_y](size_type i) {
+            return x[i * stride_x] * y[i * stride_y];
+        },
+        [](const ValueType &x, const ValueType &y) { return x + y; }, item_ct1,
+        tmp_work);
+}
+
+template <std::uint32_t cfg = KCFG_1D::encode(256, 32), typename ValueType>
+void compute_partial_dot(dim3 grid, dim3 block, size_t dynamic_shared_memory,
+                         sycl::queue *stream, size_type num_rows,
+                         const ValueType *x, size_type stride_x,
+                         const ValueType *y, size_type stride_y,
+                         ValueType *work)
+{
+    constexpr auto wg_size = KCFG_1D::decode<0>(cfg);
+    stream->submit([&](sycl::handler &cgh) {
+        sycl::accessor<UninitializedArray<ValueType, wg_size>, 0,
+                       sycl::access::mode::read_write,
+                       sycl::access::target::local>
+            tmp_work_acc_ct1(cgh);
+
+        cgh.parallel_for(
+            sycl_nd_range(grid, block), [=](sycl::nd_item<3> item_ct1) {
+                compute_partial_dot<cfg>(
+                    num_rows, x, stride_x, y, stride_y, work, item_ct1,
+                    (UninitializedArray<ValueType, wg_size> *)
+                        tmp_work_acc_ct1.get_pointer());
```

Comment:
C-style cast? Also the same for all later ones.

Reply:
dpct did, I think the get_pointer will not return the actual type pointer directly?

Reply:
use `tmp_work_acc_ct1.get_pointer().get()` currently, which return `pointer_t` and it can be found in line 187 of https://intel.github.io/llvm-docs/doxygen/access_8hpp_source.html